lib/matplotlib/pyplot.py

```diff
@@ -1610,9 +1659,13 @@ def subplot_mosaic(
     empty_sentinel: Any = '.',
     subplot_kw: dict[str, Any] | None = None,
     gridspec_kw: dict[str, Any] | None = None,
-    per_subplot_kw: dict[Hashable, dict[str, Any]] | None = None,
-    **fig_kw
-) -> tuple[Figure, dict[Hashable, matplotlib.axes.Axes]]:
+    per_subplot_kw: dict[str | tuple[str, ...], dict[str, Any]] |
+                    dict[_T | tuple[_T, ...], dict[str, Any]] |
+                    dict[Hashable | tuple[Hashable, ...], dict[str, Any]] | None = None,
+    **fig_kw: Any
+) -> tuple[Figure, dict[str, matplotlib.axes.Axes]] | \
+     tuple[Figure, dict[_T, matplotlib.axes.Axes]] | \
+     tuple[Figure, dict[Hashable, matplotlib.axes.Axes]]:
```

Comment:
Type hints can be removed from the implementation signature with overloads


they just clutter everything up trying to encompass all possible cases, and don't actually add anything that is not covered by the extant type hints



The examples from [PEP 484](https://peps.python.org/pep-0484/#function-method-overloading) and the [typing module docs](https://docs.python.org/3/library/typing.html#typing.overload) omit the type hints on the actual implementation.

Reply:
(This _may_ allow getting rid of the type ignores? not sure)

Reply:
If I drop all the types on the function, then `mypy --strict lib/matplotlib` gives:
```
lib/matplotlib/pyplot.py:1652: error: Function is missing a type annotation  [no-untyped-def]
```
and all the type ignores are no longer needed, but that's because the function's untyped, no? It is fine without `--strict` though.

Reply:
That is annoying, I don't wholly agree with that, looks like the reasoning is that it won't type check the implementation then, but I'm not _too_ concerned about that...

I could go either way.

Reply:
I think it can be improved by removing the `Hashable` path (and making the type var have `bound=Hashable` in there... may actually be able to get rid of `str` from all but the mosaic arg...


Actually its not clear to me what `HashableList[Hashable]` is doing for us at all, shouldn't that just be the same as `HashableList`, just not having the typevar bound to any particular type?